Issue 475 reply bug

creviera opened the pull request:
# Description

Fixes TypeError bug (partial fix for https://github.com/freedomofpress/securedrop-client/issues/475)

# Test Plan

1. Run regression test on `master` branch: `TESTS=tests/api_jobs/test_uploads.py::test_send_reply_failure_when_repr_is_none make test` see it fail
2. Run regression test on `issue-475-reply-bug` branch and see it pass
3. Send a reply in Qubes and you'll no longer see the error log: 
```
ERROR: Job <securedrop_client.api_jobs.uploads.SendReplyJob object at 0x65ed9994dc18> raised an exception: SendReplyJobException: Failed to send reply for source 31838d94-872c-496f-8828-0af5ae44c6bb due to TypeError("Can't convert 'NoneType' object to str implicitly",)
```
You will see this underlying error, which will be fixed in a follow-up PR: 
```
gpg: [stdin]: encryption failed: No public key
```

# Checklist

If these changes modify code paths involving cryptography, the opening of files in VMs, network (via the RPC service) traffic, or fine tuning of the graphical user interface, Qubes testing is required. Please check as applicable:

 - [x] I have tested these changes in Qubes
 - [ ] I do not have a Qubes OS workstation (the reviewer will need to test these changes in Qubes)

emkll commented:
Thanks @creviera I can confirm this partial fix works with securedrop-sdk==0.0.10, when running this branch locally via run.sh.

However, building the deb and running from the install package still results in a typeError as described in the initial ticket: upon investigation this is because requirements.txt uses securedrop-sdk=0.0.11, whereas dev-requirements.txt specify securedrop-sdk==0.0.10.

I think it might make sense to resolve this version divergence as part of this PR, provided the bug is not with the sdk.

creviera commented:
Ah, I see that we're still getting a TypeError when the journalist key is missing from the client. This is a different issue where we are passing None to `gpg`. I will fix this in this PR. Seems like it happens for both versions of the sdk (it's unrelated to the sdk).

emkll commented:
Perhaps related to the gpg issue you decribe above, there's an error in the client stating that there's no config.json file found. This I think contains the Journalist key fingerprint, and might be a possible reason why the replies aren't getting encrypted.

```
2019-07-17 11:07:50,280 - securedrop_client.config:27(from_home_dir) WARNING: Error opening config file at /home/user/.securedrop_client/config.json: [Errno 2] No such file or directory: '/home/user/.securedrop_client/config.json'
```

creviera commented:
@emkll I added a change that throws CryptoError when there is either a missing journalist fingerprint or source fingerprint instead of allowing None values to be passed to gpg. This fixes the TypeError we were seeing when passing keys to `gpg --encrypt` command string. 

I also added a couple more tests to make sure we can gracefully handle the case where a workstation has a missing config.json file or missing journalist key and the case where a workstation's sd-gpg vm is missing a source fingerprint.

creviera commented:
resolves https://github.com/freedomofpress/securedrop-client/issues/480
